lib/modules/datasource/hackage/index.ts

```diff
@@ -0,0 +1,52 @@
+import { joinUrlParts } from '../../../util/url';
+import * as pvpVersioning from '../../versioning/pvp';
+import { Datasource } from '../datasource';
+import type { GetReleasesConfig, Release, ReleaseResult } from '../types';
+
+export class HackageDatasource extends Datasource {
+  static readonly id = 'hackage';
+
+  constructor() {
+    super(HackageDatasource.id);
+  }
+
+  override readonly defaultVersioning = pvpVersioning.id;
+  override readonly customRegistrySupport = false;
+  override readonly defaultRegistryUrls = ['https://hackage.haskell.org/'];
+
+  async getReleases(config: GetReleasesConfig): Promise<ReleaseResult | null> {
+    // istanbul ignore if
+    if (typeof config.registryUrl === 'undefined') {
+      return null;
+    }
+    const massagedPackageName = encodeURIComponent(config.packageName);
+    const url = joinUrlParts(
+      config.registryUrl,
+      `package/${massagedPackageName}.json`,
+    );
+    const res = await this.http.getJson(url);
+    const keys = Object.keys(res.body as object);
+    return {
+      releases: keys.map((version) =>
+        versionToRelease(version, config.packageName),
+      ),
+    };
+  }
+}
+
+export function versionToRelease(
+  version: string,
+  packageName: string,
+): Release {
+  return {
+    version,
+    releaseTimestamp: null,
+    isStable: true,
+    changelogUrl:
+      'https://hackage.haskell.org/package/' +
+      packageName +
+      '-' +
+      version +
+      '/changelog',
```

Comment:
use `joinUrlParts` instead. 

Is there a source url instead? 
If yes, we can print the changelog in the PR body. 
This will only add a link to the changelog in the PR

Reply:
By source URL, do you mean the markdown? There is, e.g. [for base](https://hackage.haskell.org/package/base-4.20.0.1/src/changelog.md). But the location and casing is not uniform, so I don't think we'd able to guess it.

Reply:
It is available, by appending `/src`. I have pushed a commit with this, and all the other fixes.